libtenzir/src/tql2/parser.cpp

```diff
@@ -641,8 +640,109 @@ class parser {
+              codepoint |= (c - '0');
+            } else if (c >= 'a' && c <= 'f') {
+              codepoint |= (c - 'a' + 10);
+            } else if (c >= 'A' && c <= 'F') {
+              codepoint |= (c - 'A' + 10);
+            } else {
+              valid = false;
+              break;
+            }
+          }
+          if (not valid) {
+            diagnostic::error("invalid unicode escape sequence")
+              .primary(token.location.subloc(it - f - digits, digits + 2))
+              .throw_();
+          }
+          // Encode codepoint as UTF-8
+          if (codepoint <= 0x7F) {
+            result.push_back(static_cast<char>(codepoint));
+          } else if (codepoint <= 0x7FF) {
+            result.push_back(
+              static_cast<char>(0xC0 | ((codepoint >> 6) & 0x1F)));
+            result.push_back(static_cast<char>(0x80 | (codepoint & 0x3F)));
+          } else if (codepoint <= 0xFFFF) {
+            result.push_back(
+              static_cast<char>(0xE0 | ((codepoint >> 12) & 0x0F)));
+            result.push_back(
+              static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F)));
+            result.push_back(static_cast<char>(0x80 | (codepoint & 0x3F)));
+          } else if (codepoint <= 0x10FFFF) {
+            result.push_back(
+              static_cast<char>(0xF0 | ((codepoint >> 18) & 0x07)));
+            result.push_back(
+              static_cast<char>(0x80 | ((codepoint >> 12) & 0x3F)));
+            result.push_back(
+              static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F)));
+            result.push_back(static_cast<char>(0x80 | (codepoint & 0x3F)));
+          } else {
+            diagnostic::error("unicode codepoint out of range")
+              .primary(token.location.subloc(it - f - digits, digits + 2))
+              .throw_();
+          }
+        } else if (x == 'x') {
+          // Hexadecimal byte escape: \xHH
+          if (std::distance(it, e) < 2) {
+            diagnostic::error("incomplete hex escape sequence")
+              .primary(token.location.subloc(it - f, 4))
+              .throw_();
+          }
+          ++it;
+          auto c1 = static_cast<unsigned char>(*it);
+          ++it;
+          auto c2 = static_cast<unsigned char>(*it);
+          auto hex_to_int = [](unsigned char c) -> int {
+            if (c >= '0' && c <= '9') {
+              return c - '0';
+            }
+            if (c >= 'a' && c <= 'f') {
+              return c - 'a' + 10;
+            }
+            if (c >= 'A' && c <= 'F') {
+              return c - 'A' + 10;
+            }
+            return -1;
+          };
+          auto hi = hex_to_int(c1);
+          auto lo = hex_to_int(c2);
+          if (hi == -1 || lo == -1) {
+            diagnostic::error("invalid hex escape sequence")
+              .primary(token.location.subloc(it - f - 2, 4))
+              .throw_();
+          }
+          result.push_back(static_cast<char>((hi << 4) | lo));
```

Comment:
Can't see a check for unicode validity here.

Reply:
There's a general UTF8 validity check below for the entirety of the string. I don't need an additional one here, or do I?

Reply:
Ah, then that was dead code before. You could add something that detects it earlier, but that sounds annoying to write (or vibe), so maybe we just keep this as-is.